### PR TITLE
Improve Muon Analysis 2 single and simultaneous fitting

### LINKED ISSUE #27746: Muon Analysis 2 single and simultaneous fitting
The users of the Muon Analysis 2 GUI have reported that the fitting tab is difficult to use for their most common workflows.

This issue is to improve the functionality of the fitting tab within this GUI for single and simultaneous fits. Sequential fitting with be addressed in a separate issue.

### PR DESCRIPTION
**Pull request by StephenSmith25**
**Description of work.**
This PR is the first part of the planned improvements to the Muon Analysis fitting tab. The aim of the PR is make single, and simultaneous fits easier to undertake, especially for the users most common fitting requirements. 

Without the second part of fitting completed this PR represents a regression with regards to features, as sequential fitting has now been removed and will be re-added in the next PR. Therefore, the PR will be merged into a feature branch.

**To test:**
Verify the common workflow works as expected:
(1) Open the Muon Analysis 2 GUI
(2) Load in data, e.g MUSR 62260
(3) Go to fitting tab
(4) Add a fit function, select a workspace to do the fit over in the display workspace box
(5) Perform fit
(6) Tick the simultaneous box
(7) Cycle over the simultaneous fitting options and perform new fits, ensuring everything is okay.

More generally, make sure that the existing functionality within the GUI is not broken, and works as expected.

Fixes #27746. 

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AnthonyLim23** [comment]
I have just tested it and found a bug:

1. In the fitting tab change the fitting range and do a fit. The fit ignores the range (check the workspace)
2. When simultaneous mode is not on, I can still click the ```...``` button to get all of the parameters - not sure if we need to do anything about this
3. I added a function did a fit. Then clicked TF asymmetry mode and clicked fit - got a pop up about invalid properties
4. Changing the state of fit to raw data will change the selected workspace for the fit (the dropdown menu)

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py [hunk 1]
```diff
@@ -508,13 +429,132 @@ def clear_fit_information(self):
...
 
-    @property
-    def start_x(self):
-        return self._start_x
+    def update_selected_workspace_guess(self):
+        if self.view.is_simul_fit():
+            self.update_fit_specifier_list()
+        if self.manual_selection_made:
+            guess_selection = self.selected_data
+            self.selected_data = guess_selection
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py]: this seems circular, or am I missing something?

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py [hunk 2]
```diff
@@ -508,13 +429,132 @@ def clear_fit_information(self):
...
+        runs = 'All'
+        groups_and_pairs = self._get_selected_groups_and_pairs()
+        if self.view.is_simul_fit():
+            if self.view.simultaneous_fit_by == "Run":
+                runs = self.view.simultaneous_fit_by_specifier
+            elif self.view.simultaneous_fit_by == "Group/Pair":
+                groups_and_pairs = [self.view.simultaneous_fit_by_specifier]
+
+        guess_selection = []
+        for grppair in groups_and_pairs:
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py]: I dont know what `grppair` is supposed to be

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py [hunk 3]
```diff
@@ -508,13 +429,132 @@ def clear_fit_information(self):
...
+        guess_selection = list(set(self._check_data_exists(guess_selection)))
+
+        if len(guess_selection) > 1:  # sort the list to preserve order
+            guess_selection.sort(key=self._workspace_list_sorter)
+
+        return guess_selection
+
+    def update_fit_specifier_list(self):
+        if self.view.simultaneous_fit_by == "Run":
+            flattened_run_list = [str(item) for sublist in self.context.data_context.current_runs for item in sublist]
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py]: this is not very clear to me, due to the two for statements

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py [hunk 4]
```diff
@@ -557,29 +597,51 @@ def update_view_from_model(self, workspace_removed=None):
...
-            multi_domain_function.add(func)
-            multi_domain_function.setDomainIndex(index, index)
-
-        return multi_domain_function
-
     def _get_selected_groups_and_pairs(self):
         return self.context.group_pair_context.selected_groups + self.context.group_pair_context.selected_pairs
 
-    def _check_data_exists(self, guess_selection):
+    def _get_run_number_from_workspace(self, workspace_name):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py]: these should be the workspace naming file, as they are to do with naming. We have a method for getting the group or pair from the workspace name https://github.com/mantidproject/mantid/blob/master/scripts/Muon/GUI/Common/ADSHandler/workspace_naming.py#L60

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py [hunk 5]
```diff
@@ -557,29 +597,51 @@ def update_view_from_model(self, workspace_removed=None):
...
+        return run[0]
+
+    # the string following either 'Pair Asym; %s' or  Group; "
+    def _get_group_or_pair_from_workspace_name(self, workspace_name):
+        for grppair in self.context.group_pair_context.selected_groups + self.context.group_pair_context.selected_pairs:
+            grp = re.findall(r'%s' % grppair, workspace_name)
+            if len(grp) > 0:
+                return grp[0]
+
+    def _get_integer_for_grp_or_pair(self, workspace_name):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py]: I don't understand the purpose of this function. Also why does it return a double and not an int?

Comment (StephenSmith25) [reply scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py]: The intention here is to map the group or pair names to a number, which combined with the run number can be used to sort the workspace list in a consistent manner. I agree this is poorly named, as it now returns a double in the interval (0,1). I'll rename it.

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py (hunk 2 above)
Comment (StephenSmith25) [reply scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py]: The intention was this line is read as, "for group_or_pair in groups_and_pairs". I'll be more clearer with the naming though and change this.

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py (hunk 3 above)
Comment (StephenSmith25) [reply scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py]: The run list in current runs returns a list of lists. E.g [[62260]]. So this needs two for loops to be able to extract the runs.

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py (hunk 1 above)
Comment (StephenSmith25) [reply scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py]: Yeah, honestly not sure what the intention here was. I'll have a go at removing it.

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py (hunk 3 above)
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_presenter.py]: can we make it clearer thats what you are doing? Maybe just add a comment along the line of what you just said

**Comment by AnthonyLim23** [comment]
I tried to do a single fit and mantid crashed

**Comment by AnthonyLim23** [comment]
In FDA, load MUSR 62260
Go to transform and do an FFT and then a MaxEnt
Go to fitting
Change the plot type to Re
Check that the data selector defaults to the Re (it does)
Change the plot type to Im
Check that the data selector defaults to the Im(it does)
Change the plot type to MaxEnt
An error appears in the logger
```
Traceback (most recent call last):
  File "C:\Users\BTR75544\work\dev\mantid\scripts\Muon\GUI\Common\plotting_widget\plotting_widget_presenter.py", line 166, in handle_plot_type_changed
    self.plot_all_selected_workspaces()
  File "C:\Users\BTR75544\work\dev\mantid\scripts\Muon\GUI\Common\plotting_widget\plotting_widget_presenter.py", line 408, in plot_all_selected_workspaces
    label = self.get_workspace_legend_label(workspace)
  File "C:\Users\BTR75544\work\dev\mantid\scripts\Muon\GUI\Common\plotting_widget\plotting_widget_presenter.py", line 487, in get_workspace_legend_label
    '; ' + self._get_group_or_pair_from_workspace_name(workspace_name)
TypeError: cannot concatenate 'str' and 'NoneType' objects
```

**Review by AnthonyLim23 — APPROVED** [review]
:shipit:

we will need to update the docs as part of #27839

**Pull request merged**